File: backend/app/services/match_service.py
"""
Match service for loading and processing match data.
"""
import pandas as pd
import json
import os
from typing import List, Optional
import sys

# Add parent directory to path to import matchData
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
from matchData import match_data


def load_match_file(file_path: str) -> pd.DataFrame:
    """
    Load a single match JSON file and convert to DataFrame.
    
    Args:
        file_path: Path to JSON file
        
    Returns:
        DataFrame with match data
    """
    with open(file_path, "r", encoding="utf-8") as f:
        data = json.load(f)
    return match_data(data)


def load_all_matches(base_dir: Optional[str] = None, num_matches: int = 10) -> pd.DataFrame:
    """
    Load all match files from the backend directory.
    
    Args:
        base_dir: Base directory path (defaults to backend directory)
        num_matches: Maximum number of matches to load
        
    Returns:
        Combined DataFrame with all match data
    """
    if base_dir is None:
        # Default to backend directory (parent of app directory)
        # Get the directory where this file is located (services/)
        current_file_dir = os.path.dirname(os.path.abspath(__file__))
        # Go up to app/
        app_dir = os.path.dirname(current_file_dir)
        # Go up to backend/
        base_dir = os.path.dirname(app_dir)
    
    dfs = []
    files_found = []
    files_not_found = []
    
    for i in range(1, num_matches + 1):
        fname = f"response{i}.json"
        file_path = os.path.join(base_dir, fname)
        if os.path.exists(file_path):
            try:
                df = load_match_file(file_path)
                dfs.append(df)
                files_found.append(fname)
            except Exception as e:
                print(f"Error loading {fname}: {e}")
                continue
        else:
            files_not_found.append(fname)
    
    if not dfs:
        # Provide more helpful error message with debugging info
        error_msg = f"No match data found! Searched in: {base_dir}\n"
        error_msg += f"Files found: {files_found}\n"
        error_msg += f"Files not found: {files_not_found[:5]}...\n"
        error_msg += f"Current working directory: {os.getcwd()}\n"
        error_msg += f"Files in base_dir: {os.listdir(base_dir)[:10] if os.path.exists(base_dir) else 'Directory does not exist'}"
        raise ValueError(error_msg)
    
    df_all = pd.concat(dfs, ignore_index=True)
    df_all = df_all.sort_values(["player_id", "match_date"])
    return df_all


def get_last_match_stats(df: pd.DataFrame, player_id: Optional[str] = None) -> dict:
    """
    Get statistics from the last match.
    
    Args:
        df: DataFrame with match data
        player_id: Optional player ID to filter. If None, returns all players
        
    Returns:
        Dictionary with last match statistics
    """
    if player_id:
        player_data = df[df["player_id"] == player_id]
        if player_data.empty:
            return {"error": "Player not found"}
        last_match = player_data.iloc[-1]
    else:
        # Get the most recent match date across all players
        last_date = df["match_date"].max()
        last_match_data = df[df["match_date"] == last_date]
        # Return first player's data (or could return all)
        last_match = last_match_data.iloc[0]
    
    return {
        "player_id": last_match["player_id"],
        "hs_rate": round(last_match["hs_rate"], 2),
        "total_kills": int(last_match["total_kills"]),
        "total_shots": int(last_match["total_shots"]),
        "hs_count": int(last_match["hs_count"]),
        "match_date": last_match["match_date"].isoformat() if hasattr(last_match["match_date"], 'isoformat') else str(last_match["match_date"])
    }


def get_last_n_matches(df: pd.DataFrame, player_id: str, n: int = 5, start_date: Optional[pd.Timestamp] = None, end_date: Optional[pd.Timestamp] = None) -> List[dict]:
    """
    Get last N matches for a player, optionally filtered by date range.
    
    Args:
        df: DataFrame with match data
        player_id: Player UUID
        n: Number of matches to return
        start_date: Optional start date for filtering
        end_date: Optional end date for filtering
        
    Returns:
        List of match dictionaries
    """
    player_data = df[df["player_id"] == player_id].copy()
    if player_data.empty:
        return []
    
    # Filter by date range if provided
    if start_date is not None:
        player_data = player_data[player_data["match_date"] >= start_date]
    if end_date is not None:
        player_data = player_data[player_data["match_date"] <= end_date]
    
    if player_data.empty:
        return []
    
    # Get last N matches (sorted by date, most recent first)
    player_data = player_data.sort_values("match_date", ascending=False)
    last_n = player_data.head(n)
    
    matches = []
    for _, row in last_n.iterrows():
        matches.append({
            "hs_rate": round(row["hs_rate"], 2),
            "total_kills": int(row["total_kills"]),
            "total_shots": int(row["total_shots"]),
            "hs_count": int(row["hs_count"]),
            "match_date": row["match_date"].isoformat() if hasattr(row["match_date"], 'isoformat') else str(row["match_date"])
        })
    
    return matches


def get_average_hs_rate(df: pd.DataFrame, player_id: str, start_date: Optional[pd.Timestamp] = None, end_date: Optional[pd.Timestamp] = None) -> float:
    """
    Get average headshot rate for a player within a date range.
    
    Args:
        df: DataFrame with match data
        player_id: Player UUID
        start_date: Optional start date for filtering
        end_date: Optional end date for filtering
        
    Returns:
        Average headshot rate
    """
    player_data = df[df["player_id"] == player_id].copy()
    if player_data.empty:
        return 0.0
    
    # Filter by date range if provided
    if start_date is not None:
        player_data = player_data[player_data["match_date"] >= start_date]
    if end_date is not None:
        player_data = player_data[player_data["match_date"] <= end_date]
    
    if player_data.empty:
        return 0.0
    
    return round(player_data["hs_rate"].mean(), 2)

File: backend-node/src/services/matchService.js
/**
 * Match service for loading and processing match data.
 */
const fs = require('fs');
const path = require('path');
const { matchData } = require('../matchData');

/**
 * Load a single match JSON file and convert to player stats.
 * @param {string} filePath - Path to JSON file
 * @returns {Array<object>} Array of player statistics
 */
function loadMatchFile(filePath) {
  const data = JSON.parse(fs.readFileSync(filePath, 'utf-8'));
  return matchData(data);
}

/**
 * Load all match files from the backend directory.
 * @param {string} [baseDir] - Base directory path (defaults to backend directory)
 * @param {number} [numMatches=10] - Maximum number of matches to load
 * @returns {Array<object>} Combined array with all match data
 */
function loadAllMatches(baseDir = null, numMatches = 10) {
  if (!baseDir) {
    // Default to backend directory (where response*.json files are)
    baseDir = path.join(__dirname, '..', '..');
  }

  const allPlayers = [];
  const filesFound = [];
  const filesNotFound = [];

  for (let i = 1; i <= numMatches; i++) {
    const fname = `response${i}.json`;
    const filePath = path.join(baseDir, fname);
    
    if (fs.existsSync(filePath)) {
      try {
        const players = loadMatchFile(filePath);
        allPlayers.push(...players);
        filesFound.push(fname);
      } catch (error) {
        console.error(`Error loading ${fname}:`, error.message);
      }
    } else {
      filesNotFound.push(fname);
    }
  }

  if (allPlayers.length === 0) {
    throw new Error(
      `No match data found! Searched in: ${baseDir}\n` +
      `Files found: ${filesFound.join(', ')}\n` +
      `Files not found: ${filesNotFound.slice(0, 5).join(', ')}...`
    );
  }

  // Sort by player_id and match_date
  allPlayers.sort((a, b) => {
    if (a.playerId < b.playerId) return -1;
    if (a.playerId > b.playerId) return 1;
    return a.matchDate - b.matchDate;
  });

  return allPlayers;
}

/**
 * Get statistics from the last match.
 * @param {Array<object>} matches - Array of match data
 * @param {string} [playerId] - Optional player ID to filter
 * @returns {object} Last match statistics
 */
function getLastMatchStats(matches, playerId = null) {
  if (playerId) {
    const playerData = matches.filter(m => m.playerId === playerId);
    if (playerData.length === 0) {
      return { error: 'Player not found' };
    }
    const lastMatch = playerData[playerData.length - 1];
    return formatMatchStats(lastMatch);
  }

  // Get the most recent match date across all players
  const lastDate = Math.max(...matches.map(m => m.matchDate.getTime()));
  const lastMatchData = matches.filter(m => m.matchDate.getTime() === lastDate);
  
  if (lastMatchData.length === 0) {
    return { error: 'No matches found' };
  }

  return formatMatchStats(lastMatchData[0]);
}

/**
 * Format match statistics for API response.
 * @param {object} match - Match data
 * @returns {object} Formatted statistics
 */
function formatMatchStats(match) {
  return {
    player_id: match.playerId,
    hs_rate: Math.round(match.hsRate * 100) / 100,
    total_kills: match.totalKills,
    total_shots: match.totalShots,
    hs_count: match.hsCount,
    match_date: match.matchDate.toISOString(),
  };
}

/**
 * Get last N matches for a player.
 * @param {Array<object>} matches - Array of match data
 * @param {string} playerId - Player UUID
 * @param {number} [n=5] - Number of matches to return
 * @param {Date} [startDate] - Optional start date for filtering
 * @param {Date} [endDate] - Optional end date for filtering
 * @returns {Array<object>} List of match dictionaries
 */
function getLastNMatches(matches, playerId, n = 5, startDate = null, endDate = null) {
  let playerData = matches.filter(m => m.playerId === playerId);
  
  if (playerData.length === 0) {
    return [];
  }

  // Filter by date range if provided
  if (startDate) {
    playerData = playerData.filter(m => m.matchDate >= startDate);
  }
  if (endDate) {
    playerData = playerData.filter(m => m.matchDate <= endDate);
  }

  if (playerData.length === 0) {
    return [];
  }

  // Get last N matches (sorted by date, most recent first)
  playerData.sort((a, b) => b.matchDate - a.matchDate);
  const lastN = playerData.slice(0, n);

  return lastN.map(match => ({
    hs_rate: Math.round(match.hsRate * 100) / 100,
    total_kills: match.totalKills,
    total_shots: match.totalShots,
    hs_count: match.hsCount,
    match_date: match.matchDate.toISOString(),
  }));
}

/**
 * Get average headshot rate for a player within a timeframe.
 * @param {Array<object>} matches - Array of match data
 * @param {string} playerId - Player UUID
 * @param {Date} [startDate] - Optional start date for filtering
 * @param {Date} [endDate] - Optional end date for filtering
 * @returns {number|null} Average headshot rate or null if no data
 */
function getAverageHsRate(matches, playerId, startDate = null, endDate = null) {
  let playerData = matches.filter(m => m.playerId === playerId);

  if (playerData.length === 0) {
    return null;
  }

  if (startDate) {
    playerData = playerData.filter(m => m.matchDate >= startDate);
  }
  if (endDate) {
    playerData = playerData.filter(m => m.matchDate <= endDate);
  }

  if (playerData.length === 0) {
    return null;
  }

  const sum = playerData.reduce((acc, m) => acc + m.hsRate, 0);
  return Math.round((sum / playerData.length) * 100) / 100;
}

module.exports = {
  loadMatchFile,
  loadAllMatches,
  getLastMatchStats,
  getLastNMatches,
  getAverageHsRate,
};
